File: 2020/2/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	fmt.Println(part1(input))
	fmt.Println(part2(input))
}

func part1(input []string) int {
	r, _ := regexp.Compile("(\\d+)-(\\d+) (\\w): (\\w+)")
	valid := 0
	for _, line := range input {
		match := r.FindStringSubmatch(line)
		a, _ := strconv.Atoi(match[1])
		b, _ := strconv.Atoi(match[2])
		n := 0
		for i := range match[4] {
			if match[4][i] == match[3][0] {
				n++
			}
		}
		if n >= a && n <= b {
			valid++
		}
	}
	return valid
}

func part2(input []string) int {
	r, _ := regexp.Compile("(\\d+)-(\\d+) (\\w): (\\w+)")
	valid := 0
	for _, line := range input {
		match := r.FindStringSubmatch(line)
		a, _ := strconv.Atoi(match[1])
		b, _ := strconv.Atoi(match[2])
		if (match[4][a-1] == match[3][0] && match[4][b-1] != match[3][0]) || (match[4][a-1] != match[3][0] && match[4][b-1] == match[3][0]) {
			valid++
		}
	}
	return valid

}

File: 2020/2/solution.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.join(__dirname, 'input.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
}).split('\n');

let part1 = 0;
let part2 = 0;

const linePattern = /^(\d+)-(\d+) (\w): (\w+)$/;
for (const line of input) {
  const lineMatch = line.match(linePattern);
  const a = Number(lineMatch[1]);
  const b = Number(lineMatch[2]);
  const letter = lineMatch[3];
  const password = lineMatch[4];
  const count = password.split("").reduce((a, b) => a + (b === letter), 0);
  if (count >= a && count <= b) {
    part1++;
  }
  if ((password[a - 1] === letter) ^ (password[b - 1] === letter) ) {
    part2++;
  }
}

console.log(part1);
console.log(part2);
